### DISCUSSION TITLE
How should I handle breaking changes in a dependency?

### DISCUSSION THREAD

**Discussion by nex3** [Homebrew/discussions#4584] [Tap maintenance and Homebrew development] [answered]
### Output of `brew config`

```shell
N/A
```

### Output of `brew doctor`

```shell
N/A
```

### Description of issue

I'm the maintainer of Sass, including the [sass/homebrew-sass](https://github.com/sass/homebrew-sass) tap. Sass depends on the Dart SDK, which has recently released a breaking change in version 3.0.0. New versions of Dart Sass only support Dart 3.x.

Users have recently reported an issue (https://github.com/sass/homebrew-sass/issues/55) in which Homebrew is still installing Dart 2.x, which then breaks the installation process for Sass. I have two questions about this:

1. Why is Homebrew installing an outdated version of the Dart SDK when the [dart-lang/homebrew-dart](https://github.com/dart-lang/homebrew-dart) tap we depend on has Dart 3.x as the current stable version?
2. Is there a way we can restructure our tap to avoid this? Can we add a dependency constraint on Dart to indicate to Homebrew that we only support certain versions?

**Comment by p-linnane** [comment]
Are your users taps out of date? I just tapped `sass/sass` and `dart-lang/dart` and was able to install `sass` with no issues. It pulled down v3.0.5 of `dart` when I did so.

**Reply by nex3** [reply]
I don't know. I think it's fair to assume that a healthy portion of our users will have installed previous versions of Sass from Homebrew, so if that can result in out-of-date taps, I guess my question is how do we avoid that producing confusing breakages for those users?

**Reply by p-linnane** [reply, marked as answer]
Running `brew update` is all they should need to do before upgrading. Having an older version of `sass` installed and upgrading it should be no problem. I am leaning into this being an issue of the user having an outdated `dart-lang/dart` tap given what I experienced.

**Reply by nex3** [reply]
So the expectation that users should always run `brew update` before installing a new version of a tap?

**Reply by p-linnane** [reply]
Yes. That is a standard part of using `brew`. Please see https://docs.brew.sh/FAQ#how-do-i-update-my-local-packages

**Reply by MikeMcQuaid** [reply]
`brew install` will auto-update before unless users have disabled this. If you want something you can do in your formula: you could check the `HOMEBREW_VERSION` variable or similar and fail if it's older than expected?

**Reply by MikeMcQuaid** [reply]
Or, actually, much better something like:
```ruby
raise "Your dart is too old!" if Formula["dart-lang/dart/dart"].version < "3.0.0"
```

in your formula's `def install` method

**Reply by carlocab** [reply]
> So the expectation that users should always run `brew update` before installing a new version of a tap?

No, `brew install` should run `brew update` automatically. The only exception is when a user has set `HOMEBREW_NO_AUTO_UPDATE`. Users who set this should `brew update` before doing `brew install`. (Or, at least, they should `brew update` and try whatever they were having problems with again before filing an issue.)

**Reply by p-linnane** [reply]
I chose to call out `brew update` specifically since it definitely seemed the tap was out of date despite running `brew install`, which turned out to be correct but for a different reason.

[Quote from the user this affected](https://github.com/sass/homebrew-sass/issues/55#issuecomment-1600044551):
>Thank you so much for the help! It was an out of date tap; the tap couldn't update because the master branch got renamed to main and my homebrew wasn't getting updates for it.
>
>`brew tap --repair` solved it!
